### PR TITLE
Proposal for improved error handling

### PR DESCRIPTION
**Pull request by powelleric**
Please see the markdown file.  Happy to explain in further detail.

### CODE REVIEW

**Review by henriqueribeiro — COMMENTED** [review]
Thanks for the PR. 
The idea is great and I totally agree with this change, however we need to re-think it a bit. Comment inline.

**Comment by geertvandeweyer** [comment]
I agree with the idea of transferring std* on failure, but I would track missing files during delocalisation. Otherwise we have to cycle outputs twice in the scala code.I'll see if I can implement it

**Comment by geertvandeweyer** [comment]
also, I always check the logs through Batch and cloudwatch.  they contain logs even on delocalization failure.

**Comment by powelleric** [comment]
> also, I always check the logs through Batch and cloudwatch. they contain logs even on delocalization failure.

This is an important point. I am also getting the information from CloudWatch.

My problem there is that I only have the link to the logs via the Batch job, and the Batch job disappears after 7 days. Then, although the logs may exist, I'm not able to locate them. For that reason, our logs are set to auto delete after 7 days. 

Do you have any way of finding the logs after 7 days?

**Comment by geertvandeweyer** [comment]
yeah, you're right, the jobs disappear, making logs unfindable.  I've revised the delocalization scripts as proposed, but keep in mind that errors at this stage will never be visible in the stdout/stderr files, but only in the cloudwatch logs... 

@henriqueribeiro  : I don't see you comments as well, can you explain before I commit. I might have missed some aspects you thought about

**Comment by powelleric** [comment]
I have another proposal, unrelated to the current discussion.  I would like it if Cromwell tagged the Batch job with the Cromwell workflow id when submitting the job definition.  This would allow the tracking of the cost of the Cromwell job (by including that tag in the Cost & Usage Reports (CUR)).

It would be ideal if the value of tag included 4 values: the workflow id, task name, shard, and retry attempt.  (I believe these four values uniquely identify a Batch job.)  For example, each job could be tagged <workflow_id>-<task_name>-S<shard_num>-R<retry_num>.  Then, using the key of "CromwellJob", a tag could look like:  `CromwellJob: f2889b4c-b1ed-4709-b2b9-daf133de6291-myAlignment-S1-R1`.

I don't believe it costs anything to tag a job, but if it does, maybe tagging could be optionally enabled by the user using a new cromwell config property, like "awsTagEnable".

I am requesting this because I haven't found any way of tracking costs otherwise.  I enabled [split cost reporting](https://aws.amazon.com/about-aws/whats-new/2023/04/aws-split-cost-allocation-data-amazon-ecs-batch/) but it wasn't helpful.  I spoke to my AWS rep & a Batch specialist and they offered [this dashboard](https://github.com/aws-samples/aws-batch-operational-dashboards/blob/feature/job-cost-tracking/README.md), but it seems so complex and I think I would still have to dig into the internals to get a simple 2-column CSV table with Batch job id and cost.  Having the tags in the CUR report seems like the ultimate solution.  I'd love to hear your thoughts on the proposal and on how you approach cost tracking yourself. 

Thanks a bunch!

**Comment by geertvandeweyer** [comment]
I was just implementing tagging. I now have a tag with wfid, and one with
taskid. The latter hills the info you proposed I believe,  but I'll check.
Two tags would allow high level and detailed grouping.

Reheating to costs, does the cost explorer get all costs of a batch job?
The underlying ec2 instance and disks etc cannot be tagged from batch

https://trooper.be/merelswereld

**Comment by geertvandeweyer** [comment]
@powelleric : this is what the tagging looks like now : 

image

I added three tags : 
- workflow ID (all tasks running in one workflow)
- task ID : individual tasks, this is <wf_name>-<task_name>-<shard>-<attempt>
- workflow name : this allows grouping tasks over multipel wf's of the same type.

You'll have to update the EC2 role of cromwell though, as tagging is not included in the default permissions. 

My Pull Request is coming soon :-)

File: ImproveErrorHandling.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+if [ -f /tmp/scratch/importGVCFs-9-stdout.log ]; then aws s3 cp /tmp/scratch/importGVCFs-9-stdout.log s3://some-bucket/.... fi
+
+echo '*** COMPLETED DELOCALIZATION ***'
+echo '*** EXITING WITH RETURN CODE ***'
+rc=$(head -n 1 /tmp/scratch/importGVCFs-9-rc.txt)
+
+# Override successful return code if any outputs are missing
+if [ $rc = 0 && !(-f /tmp/scratch/genomicsdb.tar) ]; then
+	rc=255
+fi
```
[the remark this answers is not in the document]
Comment (geertvandeweyer) [reply ImproveErrorHandling.md:51]: image

heads up : I have the routine working in my local test version.

**Comment by powelleric** [comment]
@geertvandeweyer Thank you so much for your work on this.  It looks like those three tags would be more than sufficient for my purposes.

Regarding the cromwell-task-id, you have "None-4" at the end.  Are those two values the JOB_SCATTER_INDEX and the JOB_RETRY_ATTEMPT?  Those are two of the columns of the METADATA_ENTRY table in the cromwell database that are needed to find a given Batch job. (I called those the shard and retry num earlier in the discussion but I thought I should be more specific now.)

**Comment by powelleric** [comment]
> Reheating to costs, does the cost explorer get all costs of a batch job? The underlying ec2 instance and disks etc cannot be tagged from batch

That's an excellent question.  I thought it did, but I don't know for sure.  

As a test, two days ago I tried to tag a job that was already running, but nothing has appeared in my CUR so far.

What is your use for the tags if not for billing?

**Comment by geertvandeweyer** [comment]
> > Reheating to costs, does the cost explorer get all costs of a batch job? The underlying ec2 instance and disks etc cannot be tagged from batch
> 
> That's an excellent question. I thought it did, but I don't know for sure.
> 
> As a test, two days ago I tried to tag a job that was already running, but nothing has appeared in my CUR so far.
> 
> What is your use for the tags if not for billing?

I was also hoping for detailed usage/billing reports...

**Comment by geertvandeweyer** [comment]
> @geertvandeweyer Thank you so much for your work on this. It looks like those three tags would be more than sufficient for my purposes.
> 
> Regarding the cromwell-task-id, you have "None-4" at the end. Are those two values the JOB_SCATTER_INDEX and the JOB_RETRY_ATTEMPT? Those are two of the columns of the METADATA_ENTRY table in the cromwell database that are needed to find a given Batch job. (I called those the shard and retry num earlier in the discussion but I thought I should be more specific now.)

that's correct. as far as see in the code, None-1 reflects scatter and retry idx

**Comment by geertvandeweyer** [comment]
> > > Reheating to costs, does the cost explorer get all costs of a batch job? The underlying ec2 instance and disks etc cannot be tagged from batch
> > 
> > 
> > That's an excellent question. I thought it did, but I don't know for sure.
> > As a test, two days ago I tried to tag a job that was already running, but nothing has appeared in my CUR so far.
> > What is your use for the tags if not for billing?
> 
> I was also hoping for detailed usage/billing reports...

@powelleric , 

I've worked on the tagging some more. I've managed to tag EC2 instances and even the volumes with the wf-id, task-name and and wf-name. 


The tagging docs are available at : 
https://github.com/geertvandeweyer/cromwell/blob/develop_aws/supportedBackends/aws/src/main/scala/cromwell/backend/impl/aws/README.md#tagging-resources

Be aware that you need an extra policy rule in the "EC2InstanceRole"  used for the cromwell-server instance: 

"batch:TagResource"   to tag the job definitions.  



Also the error-handling suggestion is implemented. You can try it out from my fork : 

https://github.com/geertvandeweyer/cromwell/releases/tag/85.1-AWS

I'm still testing myself, but feel free to test & provide feedback.

**Comment by powelleric** [comment]
I tried to test it but hit some issues that might be unrelated.  I'm using the setup from [here](https://docs.opendata.aws/genomics-workflows/orchestration/cromwell/cromwell-overview.html) and I switched out the v78 jar file for the one you linked to and then restarted the cromwell-server.

In the logs, I see this error:
```
Caused by: liquibase.exception.MigrationFailedException: Migration failed for change set changesets/enlarge_call_cache_entry_ids.xml::recreate_call_caching_hash_entry_call_caching_entry_id
_fk::sshah:
     Reason: liquibase.exception.DatabaseException: Can't write; duplicate key in table '#sql-361f_939' [Failed SQL: (1022) ALTER TABLE `cromwell`.`CALL_CACHING_HASH_ENTRY` ADD CONSTRAINT
`FK_CALL_CACHING_HASH_ENTRY_CALL_CACHING_ENTRY_ID` FOREIGN KEY (`CALL_CACHING_ENTRY_ID`) REFERENCES `cromwell`.`CALL_CACHING_ENTRY` (`CALL_CACHING_ENTRY_ID`)]
```

I looked at [this script](https://github.com/geertvandeweyer/cromwell/blob/develop_aws/database/migration/src/main/resources/changesets/enlarge_call_cache_entry_ids.xml) and can't see why it wouldn't work.  I'm on an old version of MySQL from the setup I mentioned.  Could the database be the issue?  It says version `5.7.mysql_aurora.2.11.3`.

**Comment by geertvandeweyer** [comment]
I'm not familiar with the database. could it be related to the view you proposed/created above ? It seems related to  some hashing table for call caching. 

would a fresh deployment be an option? the database would be empty, but you'd have updated launch templates with improved error handling in EC2 launching. compared to the setup you linked. 

see : https://github.com/geertvandeweyer/cromwell/blob/develop_aws/supportedBackends/aws/src/main/scala/cromwell/backend/impl/aws/DEPLOY.md

**Comment by powelleric** [comment]
I was able to resolve my issue by deleting the foreign key `FK_CALL_CACHING_HASH_ENTRY_CALL_CACHING_ENTRY_ID` and restarting the cromwell-server.  It recreated the foreign key and everything seems fine.

I can confirm the tagging worked for the Batch job (see screenshot).  In a day or two, the new tags should show up 
in "AWS Billing > Cost allocation tags".  I will then enable the tags and in a few more days I will hopefully see the tags in the CUR.  This is based on the documentation found [here](https://aws.amazon.com/blogs/aws-cloud-financial-management/la-improve-cost-visibility-of-containerized-applications-with-aws-split-cost-allocation-data-for-ecs-and-batch-jobs/).  (Note that I already have the "split-cost" feature enabled on my account.)  I will report back on the status of these developments.

Thanks again for all the help!

image

**Comment by powelleric** [comment]
I see the tags in the CUR being applied to EC2 instances, including spot, ECS and data transfer.  Unfortunately, it does not seem to propagate to the EBS volumes. 

Here are the exact billing items I see tagged for one of our common runs:

|lineItem/ProductCode|lineItem/UsageType     |
|--------------------|-----------------------|
|AmazonEC2           |DataTransfer-In-Bytes  |
|AmazonEC2           |DataTransfer-Out-Bytes |
|AmazonEC2           |SpotUsage:m4.xlarge    |
|AmazonECS           |USE1-ECS-EC2-GB-Hours  |
|AmazonECS           |USE1-ECS-EC2-vCPU-Hours|

Interestingly, _some_ of the volumes are named `batch-worker-<name-of-my-compute-environment>`.  That is a bit helpful, but I can't split it up by job since they are all named the same.  Do you know if Cromwell is doing this naming?

**Comment by powelleric** [comment]
I have had cromwell-server crash on me the past two days with this error.  Could this be related to the new changes in this release?

```
[ERROR] [10/05/2023 01:03:32.281] [cromwell-system-akka.dispatchers.service-dispatcher-349] [akka.dispatch.Dispatcher] Cannot invoke "akka.actor.ActorContext.system()" because the return 
value of "akka.actor.Actor.context()" is null
java.lang.NullPointerException: Cannot invoke "akka.actor.ActorContext.system()" because the return value of "akka.actor.Actor.context()" is null
        at akka.actor.ActorLogging.log(Actor.scala:338)
```

**Comment by powelleric** [comment]
Pretty sure I've found a bug.  Here are my errors from cromwell.  I believe it is complaining about the parentheses in "Some(11)".  It would be better if it just said "11" (e.g. `sampleLevelWorkflow_WGS.baseRecalibrator-11-1`).   This is the scatter index we're talking about.  I think [this](https://github.com/geertvandeweyer/cromwell/commit/ceead50b5f8dcc6b824d8c46eae89fc678f074fe#diff-722cf69e3fe4aff5384e01cd24f3c8b766990d179c66918c7b0d7f50dcc31f29R189) is the line controlling that.

```
'sampleLevelWorkflow_WGS.baseRecalibrator-Some(11)-1' is not a valid value for TagValue - it contains illegal characters (Service: Batch, Status Code: 400, Request ID: f3a89987-f22e-4089-b888-4697663a2cfa)

software.amazon.awssdk.services.batch.model.BatchException: 'sampleLevelWorkflow_WGS.baseRecalibrator-Some(11)-1' is not a valid value for TagValue - it contains illegal characters (Service: Batch, Status Code: 400, Request ID: f3a89987-f22e-4089-b888-4697663a2cfa)
```

**Comment by powelleric** [comment]
> Pretty sure I've found a bug. Here are my errors from cromwell. I believe it is complaining about the parentheses in "Some(11)". 

@geertvandeweyer Just let me know if you need and extra details from the logs, etc regarding this potential bug.  Thanks a bunch for your help.

**Comment by geertvandeweyer** [comment]
> > Pretty sure I've found a bug. Here are my errors from cromwell. I believe it is complaining about the parentheses in "Some(11)".
> 
> @geertvandeweyer Just let me know if you need and extra details from the logs, etc regarding this potential bug. Thanks a bunch for your help.

I just ran into the same problem. it's related to the brackets in the tag value "Some(11)".   I'll look into it

**Comment by powelleric** [comment]
Is there anything I can do to help in resolving the issue with the parentheses in the tags?  I have many workflows that require this feature so I can definitely help to test that it has been resolved.

**Comment by geertvandeweyer** [comment]
I have a local build where this is fixed. but I noticed it tags with the subworkflow ids instead of main wf ids (if nested wfs are used). I tried to fix that, but couldn't find a solution yet

**Comment by powelleric** [comment]
> I have a local build where this is fixed. but I noticed it tags with the subworkflow ids instead of main wf ids (if nested wfs are used). I tried to fix that, but couldn't find a solution yet. If that's not a problem, I can push it for testing

That's definitely not an issue for us.  That would great if you could make a release with the fix.  Thank you for your help.

**Comment by geertvandeweyer** [comment]
Hi, 

I managed to extract the root workflow after all.  The release with the fix can be tried here: 

https://github.com/geertvandeweyer/cromwell/releases/download/85.1-AWS/cromwell-85-53aa961-SNAP.jar

Some minor remark : There is no prevention of race-conditions. If two tasks update the tags simultaneously, some tags might be missing. This mainly means that large instances running multiple jobs might miss "task-id" tags. 

example of instance tags : 

image

**Comment by geertvandeweyer** [comment]
@henriqueribeiro : you can merge this now if you're ok with the changes :-)

**Comment by henriqueribeiro** [comment]
@geertvandeweyer this PR only has changes on MD file. Can you please create the PR from your fork? :smile:

**Comment by geertvandeweyer** [comment]
Sorry, all the fixes are in the open pr from my fork

https://trooper.be/merelswereld

**Pull request closed without merging**